ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml

```diff
@@ -347,30 +347,42 @@ let send_host_rrd_to_master master_address =
     datasource with the name {ds_name} *)
 let add_ds ~rrdi ~ds_name =
   let open Ds in
-  try
-    let ds = List.find (fun ds -> ds.ds_name = ds_name) rrdi.dss in
+  let add ds =
     let now = Unix.gettimeofday () in
     Rrd.rrd_add_ds rrdi.rrd now
       (Rrd.ds_create ds.ds_name ds.ds_type ~mrhb:300.0 Rrd.VT_Unknown)
-  with Not_found -> raise (Rrdd_error (Datasource_missing ds_name))
+  in
+  let fail = raise (Rrdd_error (Datasource_missing ds_name)) in
+
+  List.find_opt (fun ds -> ds.ds_name = ds_name) rrdi.dss
+  |> Option.fold ~some:add ~none:fail
+
+let add_rrdi rrds uuid domid ds_name rrdi =
```

Comment:
Naming is a bit inconsistent - this could be just `add` to match the others or they could have a `_rrdi` suffix.

Reply:
I think a global `fail` function that raises `Rrdd_error (Datasource_missing fmt)` and takes a printf-style format string could be useful.

Reply:
I've changed the name to add and made sure it's not exposed to other modules by adding an mli.

I'll look into the fail one

Reply:
I've added the fail function, it does shorten the functions quite a bit